packages/vm/src/buildBlock.ts

```diff
@@ -150,20 +155,56 @@ export class BlockBuilder {
     // According to the Yellow Paper, a transaction's gas limit
     // cannot be greater than the remaining gas in the block
     const blockGasLimit = toType(this.headerData.gasLimit, TypeOutput.BigInt)
+
+    // Set default values for data gas calculations so Typescript won't complain
+    let dataGasLimit = BigInt(0)
+    let dataGasPerBlob = BigInt(0)
+
     const blockGasRemaining = blockGasLimit - this.gasUsed
     if (tx.gasLimit > blockGasRemaining) {
       throw new Error('tx has a higher gas limit than the remaining gas in the block')
     }
-
+    let excessDataGas = undefined
+    if (tx instanceof BlobEIP4844Transaction) {
+      if (this.blockOpts.common?.isActivatedEIP(4844) !== true) {
+        throw Error('eip4844 not activated yet for adding a blob transaction')
+      }
+      const blobTx = tx as BlobEIP4844Transaction
+      dataGasLimit = this.vm._common.param('gasConfig', 'maxDataGasPerBlock')
+      dataGasPerBlob = this.vm._common.param('gasConfig', 'dataGasPerBlob')
+
+      if (
+        this.dataGasUsed + BigInt(blobTx.versionedHashes.length) * dataGasPerBlob >
+        dataGasLimit
+      ) {
+        throw new Error('block data gas limit reached')
+      }
```

Comment:
Good catch.  I've updated the `validateBlobTransactions` method in `block` to include this check as well (along with a corresponding test).

Reply:
A reason to not put this in `validateBlobTransactions()` but in `validateData()` would be that `validateData()` works in a "chainless" context (with no parent block or similar needed) and can (and is) therefore be applied to more contexts.

This (`Block.validateData()`) is e.g. called in [VM.runBlock()](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/vm/src/runBlock.ts#L233) where we just can't inject/add `validateBlobTransactions()` and which would now omit this check there.

So I would still pledge to move to `validateData()`, but also open for additional arguments why it should go to `validateBlobTransactions()` I haven't thought of.

Reply:
Ah, that makes sense.  Maybe it makes the most sense to move this check inside of `validateTransactions` and do the checks alongside the 1559 checks.  Will move there since that gets called from `validateData`.